### PR TITLE
Prom stats perf improvements

### PR DESCRIPTION
**Pull request by rulex123**
Commit Message: patch to add support for streaming/chunking of prometheus stats - basically adapts the algorithm introduced in https://github.com/envoyproxy/envoy/pull/19693 for prometheus stats (while keeping common parts of the algorithm in a shared class to avoid code duplication). 
With this patch we should be able to improve the issue described here https://github.com/envoyproxy/envoy/issues/16139

Risk Level: only underlying implementation changes, but otherwise no change of behaviour/features. So low risk I would say.
Testing: current tests should remain unchanged (though they might be moved around due to refactoring of code); new tests are also added.

Docs Changes: n/a
Release Notes: n/a
[Optional Runtime guard:]
[Optional Fixes #Issue]: https://github.com/envoyproxy/envoy/issues/16139

Draft PR (already went through some review) -> https://github.com/envoyproxy/envoy/pull/24716

Benchmark data from running `stats_handler_speed_test.cc` on a dev box:

BEFORE

```
---------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations
---------------------------------------------------------------------------
BM_AllCountersText                      744 ms          744 ms            1
BM_UsedCountersText                    76.5 ms         76.4 ms            9
BM_FilteredCountersText                2866 ms         2866 ms            1
BM_Re2FilteredCountersText              371 ms          371 ms            2
BM_AllCountersJson                      917 ms          917 ms            1
BM_UsedCountersJson                    79.0 ms         79.0 ms            9
BM_FilteredCountersJson                3087 ms         3086 ms            1
BM_Re2FilteredCountersJson              369 ms          369 ms            2
BM_AllCountersPrometheus               8125 ms         8124 ms            1
BM_UsedCountersPrometheus               334 ms          334 ms            2
BM_FilteredCountersPrometheus          3680 ms         3680 ms            1
BM_Re2FilteredCountersPrometheus       1022 ms         1022 ms            1
```

AFTER

```
---------------------------------------------------------------------------
Benchmark                                 Time             CPU   Iterations
---------------------------------------------------------------------------
BM_AllCountersText                      742 ms          742 ms            1
BM_UsedCountersText                    77.6 ms         77.6 ms            9
BM_FilteredCountersText                2862 ms         2861 ms            1
BM_Re2FilteredCountersText              509 ms          508 ms            2
BM_AllCountersJson                     1559 ms         1558 ms            1
BM_UsedCountersJson                    81.0 ms         81.0 ms            9
BM_FilteredCountersJson                2857 ms         2856 ms            1
BM_Re2FilteredCountersJson              373 ms          373 ms            2
BM_AllCountersPrometheus               1842 ms         1842 ms            1
BM_UsedCountersPrometheus              92.4 ms         92.4 ms            8
BM_FilteredCountersPrometheus          2863 ms         2862 ms            1
BM_Re2FilteredCountersPrometheus        372 ms          372 ms            2
```

### CODE REVIEW

**Comment by rulex123** [comment]
@jmarantz ready for review when you are :-) Thanks!

File: source/server/admin/grouped_stats_request.cc
Comment (jmarantz) [inline source/server/admin/grouped_stats_request.cc:19]: Hm.

I suggested renaming this GroupedStatsRequest and I'm fine with sticking with it, but right now there's prometheus-specific stuff in here -- type 'type' policy, the prometheus-specific text-readouts policy, maybe other things.

Maybe just leave a TODO to clean this up if we decide to build a grouped view in HTML or JSON.

File: source/server/admin/grouped_stats_request.cc
Comment (jmarantz) [inline source/server/admin/grouped_stats_request.cc:54]: the way you wrote this I think we might get a superfluous allocation when !inserted.

WDYT of inserting an empty vector and appending the stat unconditionally?

It might be slightly faster, both because of 1 less call to the memory allocator and one less branch.

File: source/server/admin/grouped_stats_request.cc
Comment (jmarantz) [inline source/server/admin/grouped_stats_request.cc:89]: perf nit: use SymbolTable::sortByStatName, which is slightly faster because it will hold the symbol table lock over the entire sorting operation, rather than taking the and releasing its lock for each comparison.

File: source/server/admin/grouped_stats_request.cc
Comment (jmarantz) [inline source/server/admin/grouped_stats_request.cc:84]: I think you can remove the comment -- I think the code speaks for itself :)

File: source/server/admin/grouped_stats_request.cc
Comment (jmarantz) [inline source/server/admin/grouped_stats_request.cc:97]: ASSERT that it's a gauge if not a counter? It looks like you can get here with a text-readout; is that right?

This may be correct but it could use some comments.

File: source/server/admin/grouped_stats_request.cc
Comment (jmarantz) [inline source/server/admin/grouped_stats_request.cc:131]: SymbolTable::sortByStatNames

File: source/server/admin/grouped_stats_request.cc
Comment (jmarantz) [inline source/server/admin/grouped_stats_request.cc:138]: I guess this is where a Render interface might be useful. Let's just call this out as prometheus-specific and it could be refactored incrementally later.

Also, you *do* have a file of prom rendering functions -- should we at least move this magic there?

File: source/server/admin/grouped_stats_request.h
Comment (jmarantz) [inline source/server/admin/grouped_stats_request.h:18]: a class comment would be good, something like:

// Renders stats grouped by tag-extracted name. This is currently used for Prometheus (and has some
// Prometheus-specific logic in it) but this could be generalized for other formats.

File: source/server/admin/prometheus_stats_formatter.cc
Comment (jmarantz) [inline source/server/admin/prometheus_stats_formatter.cc:15]: I assume this code is unchanged from what was in prometheus_stats.cc.

Actually would it make more sense to merge this with the prometheus render code below?

**Review by jmarantz — COMMENTED** [review]
flushing some comments; sorry this has taken longer than expected to review.

**Review by jmarantz — COMMENTED** [review]
flushing a couple more comments

File: source/server/admin/prometheus_stats_formatter.cc
Comment (rulex123) [reply source/server/admin/prometheus_stats_formatter.cc:15]: Yep, this is pre-existing code: sure, I will move it to the `PrometheusStatsRender` class.

File: source/server/admin/grouped_stats_request.cc
Comment (rulex123) [reply source/server/admin/grouped_stats_request.cc:97]: Yep, you can reach this code with a text-readout, which are considered of type `gauge` when rendered (see https://github.com/envoyproxy/envoy/blob/9d9940b49f642fcf483ac8ff387ecb521dcf94fc/source/server/admin/prometheus_stats.cc#L278-L280).

I will add a comment to explain that.

File: source/server/admin/grouped_stats_request.cc
Comment (rulex123) [reply source/server/admin/grouped_stats_request.cc:54]: Hm I am not sure that will work though. Inserting an empty vector would "reset" any other stats with the same tag extracted name, but different tags

Comment (jmarantz) [reply source/server/admin/grouped_stats_request.cc:54]: From https://en.cppreference.com/w/cpp/container/map/insert

> insert(value_type): Returns a pair consisting of an iterator to the inserted element (or to the element that prevented the insertion) and a bool value set to true if and only if the insertion took place.

So it won't overwrite.

Comment (rulex123) [reply source/server/admin/grouped_stats_request.cc:54]: aha, got it! will update the code

File: source/server/admin/grouped_stats_request.cc
Comment (rulex123) [reply source/server/admin/grouped_stats_request.cc:18]: I have included a TODO in the header file.

File: source/server/admin/grouped_stats_request.cc
Comment (rulex123) [reply source/server/admin/grouped_stats_request.cc:85]: I tried swapping this with SymbolTable::sortByStatName and I am getting a couple of failures in grouped_stats_requests_test.cc with some lock-related failure (excerpt below)

```
[mutex.cc : 1391] RAW: Acquiring absl::Mutex 0x7fd91a812b28 while holding  0x7fd91a812a18; a cycle in the historical lock ordering graph has been observed
```
I am still digging, though I don't think further reviewing has to block on this.

File: source/server/admin/stats_render.h
Comment (rulex123) [inline source/server/admin/stats_render.h:102]: Perhaps it would make more sense to have `GroupedStatsRender` and `UngroupedStatsRender`, but IMO we could leave that to a future refactoring.

File: source/server/admin/grouped_stats_request.cc
Comment (rulex123) [reply source/server/admin/grouped_stats_request.cc]: Fixed now!

File: source/server/admin/stats_render.h
Comment (jmarantz) [reply source/server/admin/stats_render.h:102]: I think what you have now is good. I feel like "Prometheus" vs (say) "Html" or "Text" is one dimension, and "Grouped" vs "Ungrouped" is orthogonal.

File: test/server/admin/grouped_stats_request_test.cc
Comment (jmarantz) [inline test/server/admin/grouped_stats_request_test.cc:128]: this is great. I just want to leave a note for @ggreenway for senior maintainer review that these previous tests were moved to the prometheus section of render_test.cc, not removed.

**Review by jmarantz — APPROVED** [review]
Awesome work!

**Comment by jmarantz** [comment]
@ggreenway can you take a senior maintainer pass on this?

**Comment by ggreenway** [comment]
> @ggreenway can you take a senior maintainer pass on this?

Yep, I'll review

**Comment by ggreenway** [comment]
Sorry for the delay; the code movement isn't tracked well by github so it's somewhat difficult to review. I'll post a review today or tomorrow.

File: source/server/admin/stats_render.cc
Comment (ggreenway) [inline source/server/admin/stats_render.cc:390]: Consider using absl::StrCat for consistency with formatting above.

File: source/server/admin/stats_render.cc
Comment (ggreenway) [inline source/server/admin/stats_render.cc:377]: Why are these 4 empty? Isn't this the StatsRender interface? At least add a comment about what's going on here.

File: source/server/admin/stats_request.h
Comment (ggreenway) [inline source/server/admin/stats_request.h:45]: Should the `phase_label` be looked up from a constant table based on the `PhaseName`? Or can the label vary for the `PhaseName`?

File: source/server/admin/ungrouped_stats_request.cc
Comment (ggreenway) [inline source/server/admin/ungrouped_stats_request.cc:51]: This is the same as the prometheus filtering. Can it be pulled out to a utility/helper?

File: test/server/admin/grouped_stats_request_test.cc
Comment (ggreenway) [inline test/server/admin/grouped_stats_request_test.cc:1]: Refactoring/rewriting all the tests makes it EXTREMELY difficult to understand if there were any behavior changes to what prometheus stats were output. Can you try leave the prometheus tests as untouched as possible in this PR, and we can consider revising them later?

Also, I think it makes sense to leave a whole set of prometheus-format-specific tests.

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait

File: source/server/admin/stats_handler.cc
Comment (rulex123) [reply source/server/admin/stats_handler.cc:88]: Agreed, I will update the code accordingly.

File: source/server/admin/stats_render.cc
Comment (rulex123) [reply source/server/admin/stats_render.cc:26]: I will move this and other prom-related methods out of the anonymous namespace and to the `PrometheusStatsRender` class, so I think we leave this name unchanged.

File: source/server/admin/stats_render.cc
Comment (rulex123) [reply source/server/admin/stats_render.cc:37]: Same as for `sanitizeName`.

File: source/server/admin/stats_render.cc
Comment (rulex123) [reply source/server/admin/stats_render.cc:306]: Agreed, I will update the code to factor out any common logic.

File: source/server/admin/stats_render.cc
Comment (rulex123) [reply source/server/admin/stats_render.cc:377]: Yeah, I agree this is a bit odd. I will refactor the `StatsRender` interface so that we don't end up with these empty methods in the prometheus render.

File: source/server/admin/ungrouped_stats_request.cc
Comment (rulex123) [reply source/server/admin/ungrouped_stats_request.cc:51]: Very similar but not exactly the same: the prometheus logic avoids a `stat->name()` unless there is a filter. The original code had a comment explaining this, I will add it back.

File: test/server/admin/grouped_stats_request_test.cc
Comment (rulex123) [reply test/server/admin/grouped_stats_request_test.cc:1]: Fair point about the difficulty of reviewing due to moving all the tests. I will move the pre-existing prom related ones back to original files, hopefully that will ease code review.
 
> Also, I think it makes sense to leave a whole set of prometheus-format-specific tests.
Yeah, all of the pre-existing prometheus tests are left pretty much untouched, I only added new tests in this patch. Or maybe I am misunderstanding what you mean with this comment?

File: source/server/admin/stats_render.cc
Comment (jmarantz) [reply source/server/admin/stats_render.cc:309]: I think it's OK for 'finalize' to be empty for prom -- this is empty for 'text' mode as well. It was needed for JSON format because the structure of the way the histograms are written out makes it impossible to stream the histograms. They are to collected and then emitted all once from the finalize routine.

File: source/server/admin/stats_handler.cc
Comment (jmarantz) [inline source/server/admin/stats_handler.cc:87]: this seems a little awkward; can we change the call-site `/stats/prometheus` to set the format,  and drop the extra bool?

File: test/server/admin/prometheus_stats_test.cc
Comment (jmarantz) [inline test/server/admin/prometheus_stats_test.cc:49]: I don't think we need to access the private member vars of TestScope -- we can just use the Scope API to add these, right?

File: test/server/admin/prometheus_stats_test.cc
Comment (jmarantz) [inline test/server/admin/prometheus_stats_test.cc:102]: I think you can just do test_scope_ptr_.reset() instead of all this.

**Review by jmarantz — COMMENTED** [review]
I need to a look a bit more at the other changes in this test, but flushing comments for now.

Ideally we should have almost no changes to the test body, modulo namespace changes.

File: test/server/admin/prometheus_stats_test.cc
Comment (rulex123) [reply test/server/admin/prometheus_stats_test.cc:102]: Hm `numSymbols()` of the symbol table doesn't seem to clear though when using `test_scope_ptr_ .reset()`?

Comment (jmarantz) [reply test/server/admin/prometheus_stats_test.cc:102]: Hm. That seems surprising to me. Could you try calling symbol_table_->debugPrint() to see what's still there?

Comment (jmarantz) [reply test/server/admin/prometheus_stats_test.cc:102]: never mind I know what the problem is. For that implementation, the scopes don't really own the stats; the Store does.

You'd have to clear out the MockStore, which is annoying since it's not a shared_ptr. What you have is fine. Sorry for the confusion. You could leave a comment saying why we have to reach into the Store to clean all the stats in it, and maybe we could clean it up later.

File: test/server/admin/prometheus_stats_test.cc
Comment (rulex123) [reply test/server/admin/prometheus_stats_test.cc:49]: Possibly, but then we would have to change the `PrometheusStatsFormatterTest` class even more? 

My idea was to keep the test setup as close as possible the current version, that's why I ended up adding a new `TestScope` class and accessing its private vars. Do we need to change that?

File: test/server/admin/stats_request_test_base.h
Comment (jmarantz) [inline test/server/admin/stats_request_test_base.h:42]: I missed this before. Why do we need this version of TestScope?

Can't we just use MockStore::createScope()?

Actually come to think of it, can we just use IsolatedStoreImpl directly, or maybe Stats::TestUtil::TestStore and TestScope, without introducing these new wrappers?

File: test/server/admin/prometheus_stats_test.cc
Comment (jmarantz) [reply test/server/admin/prometheus_stats_test.cc:49]: I think it would not be more of a change that what you have already. You can just call test_scope_ptr_->counterFromStatName().

I noted this in the header file you added, but I'm not clear why you needed to subclass IsolatedStore and IsolatedScope, rather than just using them directly.

File: source/server/admin/stats_request.h
Comment (jmarantz) [inline source/server/admin/stats_request.h:111]: I think these can be in the protected section.

File: source/server/admin/ungrouped_stats_request.cc
Comment (jmarantz) [inline source/server/admin/ungrouped_stats_request.cc:72]: it would be good to brainstorm about how to get rid of the dynamic_cast. I hadn't noticed these before.

They are a mild anti-pattern IMO, and also they are slow.

They don't seem too hard to get rid of -- can't we have the `render_` object in UngroupedStatsRequest be a StatsRender?

File: source/server/admin/stats_request.h
Comment (jmarantz) [inline source/server/admin/stats_request.h:128]: Hmm. WDYT of leaving this private and moving some of these to the subclass (e.g. render_ in particular), or accessing them from subclasses via protected API?

**Review by jmarantz — COMMENTED** [review]
Great progress.

I feel like some of the recent changes require some further thought.

I am regretting suggesting that we fold Prom into the StatsRender class hierarchy; I think this might have added complexity that may only benefit us in the future if we do a grouped version of the HTML renderer. This was my bad.

I'm now thinking we should go back to keeping the Prom version of StatsRequest separate and just do its rendering in the most natural way.

File: source/server/admin/stats_request.h
Comment (rulex123) [reply source/server/admin/stats_request.h:111]: I revisited all of the access modifiers in `stats_request.h`, I realise I had unnecessarily "relaxed" some of the pre-existing access.

File: test/server/admin/stats_request_test_base.h
Comment (rulex123) [reply test/server/admin/stats_request_test_base.h:42]: I think the issue with using existing test infrastructure for scope|store boils down to histograms: the `PrometheusStatsFormatterTest` tests heavily rely on mocking behaviour for histograms and I don't see an easy way to plug that into `IsolatedScopeImpl`.

I also looked a bit at `MockScope|Store` but I don't think they can be used as is for those tests: I might be missing something here, so LMK in case I am wrong :)

File: source/server/admin/stats_request.cc
Comment (jmarantz) [inline source/server/admin/stats_request.cc:54]: perf nit: hoist the render object into a temp outside the outer `while` avoid calling the same virtual call 3x per iteration.

File: source/server/admin/ungrouped_stats_request.h
Comment (jmarantz) [inline source/server/admin/ungrouped_stats_request.h:16]: Add comment explaining that this is for treating each stat on its own, as opposed to the Grouped version which groups by tag-name.

File: test/server/admin/stats_request_test_base.h
Comment (jmarantz) [reply test/server/admin/stats_request_test_base.h:42]: Let's leave this like you have it but it's a bit of technical debt to have to subclass IsolatedScopeImpl for an individual set of tests. There are so many Scope impls we should be able to re-use one.

Can you add a TODO for me to look at this and try to think of how to clean it after this is merged?

**Review by jmarantz — COMMENTED** [review]
Mostly nits remaining. Thank you.

Greg -- there are a lot of changes now to prometheus_stats_test.cc to to calling model,, but it's not too hard to scan it and see it's testing a superset of what was previously tested. WDYT?

**Review by jmarantz — APPROVED** [review]
Thank you!

**Comment by jmarantz** [comment]
@ggreenway over to you.

**Comment by htuch** [comment]
@ggreenway friendly ping.

**Comment by adisuissa** [comment]
@ggreenway can you PTAL?

File: source/server/admin/stats_render.cc
Comment (jmarantz) [reply source/server/admin/stats_render.cc:400]: This makes sense.

FYI in https://github.com/envoyproxy/envoy/pull/26472 I have added a new bucket mode tentatively called 'detailed' which will have the raw bucket data. That *may* also be useful for Prom as you can see a lot more interesting data with this IMO, but it might complicate aggregation in Prom across multiple Envoys.

Don't worry Greg -- that PR is still a ways away from being atomized into probably around 4 for review, but I think the first PR will be the addition of a new bucket-mode to show the original buckets computed by circlhist.

Nevertheless I agree removing this TODO is fine for now.

**Review by jmarantz — APPROVED** [review]

**Review by ggreenway — APPROVED** [review]

**Pull request merged**
